lib/product-import.coffee

```diff
@@ -417,18 +420,14 @@ class ProductImport
       # if the existing products were not fetched, load them from API
       .then (existingProducts) =>
         existingProducts or @_getExistingProductsForSkus(@_extractUniqueSkus([prodToProcess]))
-      .then (existingProducts) =>
-        if (@beforeCreateOrUpdateCallback)
-          return Promise.resolve(@beforeCreateOrUpdateCallback(prodToProcess, existingProducts))
-            .then () ->
-              return existingProducts
-        else
-          return existingProducts
       .then (existingProducts) =>
         if existingProducts.length
           @_updateProductRepeater(prodToProcess, existingProducts, reassignmentRetries)
         else
           @_prepareNewProduct(prodToProcess)
+            .then (product) =>
+              Promise.resolve(@beforeCreateCallback(product))
```

Comment:
- I would pass here `prodToProcess` too. It represents a draft with keys on references which are not available on `product` anymore and it might be useful to the user
- I would avoid here mutation of parameters too and expect to get a new `product` as result instead. In case no product is returned we should not issue create request.